From 42b2d042d47390cac1b483ad9629fe73c91253ea Mon Sep 17 00:00:00 2001
From: Peter Yin <peter.yin@quantatw.com>
Date: Fri, 1 Aug 2025 18:33:49 +0800
Subject: [PATCH 22/22] pldmd: oem: Add MMC VR failure event

link:https://gerrit.openbmc.org/c/openbmc/pldm/+/82142

Testd:
'''
MMC send command
uart:~$ test pldm 8 3f 2 5 5 0 0 0 27 1 2 7a 7b
RESP
00000000: 00

BMC Log
{
"additional_data": {
"FAILURE_DATA": "VR Fault Status: 0x7a7b",
"VOLTAGE_REGULATOR": "/xyz/openbmc_project/inventory/MTIA_VR_P0V75_PVDD_CH_N",
"_CODE_FILE": "/usr/src/debug/pldm/1.0+git/oem/meta/libpldmresponder/oem_meta_file_io_type_event_log.cpp",
"_CODE_FUNC": "virtual int pldm::responder::oem_meta::EventLogHandler::write(const pldm::responder::oem_meta::message&)",
"_CODE_LINE": "197",
"_PID": "1480"
},
"event_id": "",
"message": "xyz.openbmc_project.State.Power.VoltageRegulatorFault",
"redfish": {
"args": [
"/xyz/openbmc_project/inventory/MTIA_VR_P0V75_PVDD_CH_N"
],
"id": "OpenBMC_StatePower.VoltageRegulatorFault",
"message": "A fault has been detected on voltage regulator '/xyz/openbmc_project/inventory/MTIA_VR_P0V75_PVDD_CH_N'"
},
"resolution": "",
"resolved": false,
"severity": "xyz.openbmc_project.Logging.Entry.Level.Error",
"timestamp": "2025-07-09T08:33:27.877000000Z",
"updated_timestamp": "2025-07-09T08:33:27.877000000Z"
}
---
 .../oem_meta_file_io_type_event_log.cpp       | 66 ++++++++++++++++---
 1 file changed, 58 insertions(+), 8 deletions(-)

diff --git a/oem/meta/libpldmresponder/oem_meta_file_io_type_event_log.cpp b/oem/meta/libpldmresponder/oem_meta_file_io_type_event_log.cpp
index 57a720a..c929455 100644
--- a/oem/meta/libpldmresponder/oem_meta_file_io_type_event_log.cpp
+++ b/oem/meta/libpldmresponder/oem_meta_file_io_type_event_log.cpp
@@ -182,6 +182,30 @@ int EventLogHandler::write(const message& eventData)
         {" PVPP_CD_ASIC2", false},
     });
 
+    static const auto mtia_vr_source = std::to_array<VRSource>({
+        {" MTIA_VR_P3V3", true},
+        {" MTIA_VR_P0V85_PVDD", true},
+        {" MTIA_VR_P0V75_PVDD_CH_N", true},
+        {" MTIA_VR_P0V75_MAX_PHY_N", true},
+        {" MTIA_VR_P0V75_PVDD_CH_S", true},
+        {" MTIA_VR_P0V75_MAX_PHY_S", true},
+        {" MTIA_VR_P0V75_TRVDD_ZONEA", true},
+        {" MTIA_VR_P1V8_VPP_HBM0_HBM2_HBM4", true},
+        {" MTIA_VR_P0V75_TRVDD_ZONEB", true},
+        {" MTIA_VR_P0V4_VDDQL_HBM0_HBM2_HBM4", true},
+        {" MTIA_VR_P1V1_VDDC_HBM0_HBM2_HBM4", true},
+        {" MTIA_VR_P0V75_VDDPHY_HBM0_HBM2_HBM4", true},
+        {" MTIA_VR_P0V9_TRVDD_ZONEA", true},
+        {" MTIA_VR_P1V8_VPP_HBM1_HBM3_HBM5", true},
+        {" MTIA_VR_P0V9_TRVDD_ZONEB", true},
+        {" MTIA_VR_P0V4_VDDQL_HBM1_HBM3_HBM5", true},
+        {" MTIA_VR_P1V1_VDDC_HBM1_HBM3_HBM5", true},
+        {" MTIA_VR_P0V75_VDDPHY_HBM1_HBM3_HBM5", true},
+        {" MTIA_VR_P0V8_VDDA_PCIE", true},
+        {" MTIA_VR_P1V2_VDDHTX_PCIE", true},
+        {" MTIA_VR_MAX", true},
+    });
+
     /* systemd service to kick start a target. */
     constexpr auto SYSTEMD_SERVICE = "org.freedesktop.systemd1";
     constexpr auto SYSTEMD_ROOT = "/org/freedesktop/systemd1";
@@ -204,15 +228,19 @@ int EventLogHandler::write(const message& eventData)
         return PLDM_ERROR;
     }
 
-    try
-    {
-        parseObjectPathToGetSlot(slot);
-    }
-    catch (const std::exception& e)
+    /* 1. Get host slot number */
+    if (!configurations.empty())
     {
-        slot=0;
-        warning("Using default slot0 TID={TID}, WARNING={WARNING}",
-                "TID", tid, "WARNING", e);
+        try
+        {
+            parseObjectPathToGetSlot(slot);
+        }
+        catch (const std::exception& e)
+        {
+            error("Using default slot0 TID={TID}, WARNING={WARNING}",
+                    "TID", tid, "WARNING", e);
+            return PLDM_ERROR;
+        }
     }
 
     /* 2. Map event from list */
@@ -273,6 +301,28 @@ int EventLogHandler::write(const message& eventData)
 
         event_name = eventList[eventType];
     }
+    else if (strcmp(eventList[eventType], " MTIA VR FAULT") == 0)
+    {
+        if (eventData[2] >= mtia_vr_source.size())
+        {
+            error("{FUNC}: Invalid vr source={MTIA_VR_SOURCE}", "FUNC",
+                  std::string(__func__), "MTIA_VR_SOURCE", eventData[2]);
+            return PLDM_ERROR;
+        }
+
+        const auto& vr = mtia_vr_source[eventData[2]];
+        const std::string_view baseName = eventList[eventType];
+
+        if (vr.isPMBus)
+        {
+            event_name = std::format("{} :{} status: 0x{:02x}{:02x}", baseName,
+                                     vr.netName, eventData[3], eventData[4]);
+        }
+        else
+        {
+            event_name = std::format("{} :{}", baseName, vr.netName);
+        }
+    }
     else
     {
         event_name = eventList[eventType];
-- 
2.43.0

